\leadauthor{Yangyang Li}

\title{PxBLAT: An Efficient and Ergonomics Python Binding Library for BLAT}
\shorttitle{Ergonomic Genomic Analysis with \acrshort{pxblat} }

\author[1]{Yangyang Li\orcidlink{0000-0001-8224-1067}}
% \author[2]{Second Doctor \orcidlink {000-0002-0000-0000}}
\author[1,\Letter]{Rendong Yang \orcidlink {000-0003-0000-0000}}
\affil[1]{Department of Urology, Northwestern University Feinberg School of Medicine, Chicago, IL 60611}
% \affil[2]{B Institute, Chalk Road, Blackboardville, USA}
\date{}

\maketitle

\begin{abstract}
	\acrshort{pxblat} provides an ergonomic and efficient Python binding library for \acrshort{blat}, designed to enhance the user experience and performance of genomic analysis tasks.
	By providing a pythonic interface to \acrshort{blat}, \acrshort{pxblat} simplifies the usage of \acrshort{blat}, allowing incorporating its functionality directly into their Python-based bioinformatics workflows.
	Furthermore, A new parallelization and non-blocking model enables improved efficiency and intuitive user interface, thereby reducing the complexity of genomic data analysis tasks.
\end{abstract}


\begin{keywords}
	Software Libraries |  Sequence Analysis | BLAT
\end{keywords}

\begin{corrauthor}
	rendong.yang\at northwestern.edu
\end{corrauthor}

\section*{Introduction}\label{sec:introduction}

The continual advancement of genome sequencing technologies has led to an exponential increase in available genomic data.
Tools to analyze and manipulate these data have become critically important in both research and clinical contexts.
\acrfull{blat}~\citep{kent2002blat} is a standout in the bioinformatics field for its capability to perform rapid genome sequence alignments.
It offers a faster alternative to \acrfull{blast}~\citep{altschul1990basic}  for aligning DNA sequences to the human genome ~\citep{kent2002blat}.
Despite its widespread use and acceptance in the bioinformatics community, interfacing with BLAT can present challenges, particularly when integrating it within a broader Python-based analytical pipeline.
Since \acrshort{blat} is implemented by C programming language and only provides a variety of utilities with a \acrfull{cli}.
So far, we lack a holistic approach to enhance both the performance and the usability of BLAT.

On the other hand, python's rise as a favored programming language in bioinformatics is well-documented, due to its ease of use, extensive libraries, and versatility~\citep{perkel2015programming}.
Various binding libraries have been developed to extend python's reach into other computing languages, improving the flexibility and interoperability of bioinformatics tools.
For instance, Biopython~\citep{cock2009biopython}, one of the most prominent bioinformatics libraries, provides interfaces to tools like \acrshort{blast} ~\citep{altschul1990basic}, Clustal~\citep{higgins1988clustal}, and others.
Nonetheless, to date, no comprehensive Python binding library for BLAT has been developed.

Here we propose \acrshort{pxblat}, a modern Python library designed to streamline and enhance the interaction with \acrshort{blat}, thereby making it more efficient and ergonomic.
\acrshort{pxblat} serves as a bridge, bringing the high-performance capabilities of \acrshort{blat}  into the Python environment, which is widely regarded for its readability, simplicity, and extensive library support.
By improving the usability of \acrshort{blat}  and seamless integration within Python, \acrshort{pxblat} opens up new possibilities for efficient genomic analysis.
We provide evidence of its performance improvements, demonstrate its ergonomic advantages, and discuss its potential applications in genomic research.
The overarching aim of this work is to fill the observed gap by providing a Python binding library specifically tailored for BLAT, addressing both its efficiency and ergonomic concerns.

\section*{Implementation}\label{sec:implementation}

% introduce features of pxblat
% 1. no intermidiate files, all in memory
% 2. no system call
% 3. no bother with log files to get status of tool
% 5. no need to worry about file format
% 6. no other dependency
% 4. higher proformance and Ergonomics (compare with current blat endpoint)

The design of \acrshort{pxblat} follows the pythonic principles of readability and simplicity.
It is built to be intuitive, reducing the learning curve for users familiar with Python and bioinformatics tools.
In line with this, we focused on minimizing the complexity while maximizing the usability and performance.
Hence, We employ existing C codebase and reimplement utilities of \acrshort{blat}(V37.1) suit  including \emph{faTwoBit}, \emph{gfServer} and \emph{gfClient} \acrshort{cli}  using Cpp programming language~\citep{kent2002blat}.
Then, we create \acrshort{pxblat} via Pybind11~\citep{pybind11}.
This approach enabled direct access to the functions of the \acrshort{blat} program without modifying its original source code, preserving the integrity and performance of \acrshort{blat} while extending its capabilities.

Furthermore, \acrshort{pxblat} eliminate the need for intermediate files, which allow all operations to be in memory, these tools can analyze and respond to data inputs almost instantaneously.
Input and output file now become optional, which is more flexible.
Meanwhile, \acrshort{pxblat} reduce the need of system calls to interact with \acrshort{blat}, which may prevent potential safe problem.
It provides optional non-blocking operations as well, improving the overall efficiency.
Importantly, \acrshort{pxblat} allow fetching status of \acrshort{blat} without manipulating log files, especially in concurrent environment.
That means \acrshort{pxblat} reduce possibility of data race, which is annoying issue in concurrent environment.
\acrshort{pxblat} also implement ergonomic features consisting of checking and waiting if server is ready for alignment, retrying different port if current port is occupied, using existing server if server is already started in current computing node.
All these features have been tested and work fine in concurrent environment.
Compared to \emph{gfServer},  \acrshort{pxblat} reimplement it using multiple threads model, which enables processing multiple client request concurrently.

Specifically, \acrshort{pxblat} provides \acrfull{api} including  \emph{Server}, \emph{Client} class and other free functions to reproduce \acrshort{blat} \acrshort{cli}.
\emph{Server} and \emph{Client} have same utilities as \emph{gfServer} and \emph{gfClient} respectively but more flexible and efficient.
Besides, they own ergonomic features aforementioned.
Free functions the library provides, for example, \emph{start\_server}, \emph{query\_server}, \emph{status\_server}, \emph{fa2twobit} etc., serve as potential usage under different contexts.
\acrshort{pxblat} also offers same but more user-friendly \acrshort{cli} as \acrshort{blat} suit.
The library is well tested, and use type hints for all public classes and functions to ensure the quality and correctness via static analyzer.
These type annotations also make \acrshort{pxblat} more pleasant to use inside an \acrfull{ide}, where the function signatures can be suggested and corrected automatically.

\begin{figure*}
	\centering
	\includegraphics[width=0.75\linewidth]{figures/pxblat.png}
	\caption{\textbf{These are cells.}\\
		(\textbf{A}) This is a regular figure with a legend as a caption underneath. Inset: 3X zoom. Scale bar, \SI{10}{\micro\meter}.}
	\label{fig:pxblat}
\end{figure*}


\section*{Results}\label{sec:results}

\begin{listing}
	\inputminted[linenos]{python}{codes/example1.py}
	\caption{Python example}
	\label{listing:1}
\end{listing}


To validate \acrshort{pxblat}'s utility and effectiveness, we undertook a comprehensive evaluation process, assessing the library's performance, usability, and robustness.
A series of tests and comparative analysis were carried out to demonstrate the efficiency gains and ergonomic improvements offered by \acrshort{pxblat} over \acrshort{blat} usage.
We evaluated \acrshort{pxblat}'s efficiency by comparing the execution times of identical alignment tasks performed directly using \acrshort{blat}  and via \acrshort{pxblat}.
Our tests involved datasets of varying sizes to cover a range of typical usage scenarios.
The results indicated a significant efficiency gain when using \acrshort{pxblat} (Figure~\hyperref[fig:pxblat]{1A}).
In all cases, the execution time was reduced when using \acrshort{pxblat} compared to \acrshort{blat}.
The reduction in execution time ranged from \SI{1}{\percent} to \SI{1}{\percent}, clearly demonstrating the benefits of \acrshort{pxblat} 's direct interfacing approach.
In summary, \acrshort{pxblat}  offers tangible benefits in terms of reduced execution time and improved user experience, proving its value as an enhancement to BLAT's functionality.

Through our evaluation, we have demonstrated that \acrshort{pxblat} not only accelerates the execution of BLAT operations but also presents a user-friendly,
pythonic interface that seamlessly integrates with existing Python-based bioinformatics workflows.
While the current version of \acrshort{pxblat}  has demonstrated significant advantages in terms of efficiency and ergonomics, we believe that there is potential for further development and enhancement.
\acrshort{pxblat} represents a significant contribution in this regard, offering a blend of performance and usability that stands to benefit a wide range of users.


% show length distribution of fas data and show have same results
% TODO: Benchmarking Performance <Yangyang Li>

% show performance with one thread one client model, compared to original command line version time efficiency
% threadpool can be used instead of processpool, which means do not bother wtih overhead for starting another process

% todo: mention that we can only use thread pool when doing multiple  query to accerlate our processing

\section*{Acknowledgements}\label{sec:acknowledgements}


\section*{Conflict of interest}\label{sec:conflict-of-interest}


\section*{Funding}\label{sec:funding}


\section*{Data availability}\label{sec:data-availability}

The \acrshort{pxblat}, along with the source code, example scripts, and comprehensive documentation, are publicly available in our GitHub repository at \url{https://github.com/cauliyang/pxblat}.
We have also provided extensive test data, allowing users to validate the functionality and performance of \acrshort{pxblat} in their own environments.
We welcome community contributions to the \acrshort{pxblat} project.
We believe that the open availability of our data and code will foster collaboration and iterative improvement, in line with our vision of creating accessible, efficient tools for bioinformatics.

\section*{Reference}\label{sec:reference}
\bibliographystyle{bxv_abbrvnat}
\bibliography{refs.bib}




























