\leadauthor{Yangyang Li}

\title{PxBLAT: An Ergonomic and Efficient Python Binding Library for BLAT}
\shorttitle{Ergonomic Genomic Analysis with \gls{pxblat} }

\author[1]{Yangyang Li\orcidlink{0000-0001-8224-1067}}
\author[1,\Letter]{Rendong Yang \orcidlink {000-0003-0000-0000}}
\affil[1]{Department of Urology, Northwestern University Feinberg School of Medicine, Chicago, IL 60611}
% \affil[2]{B Institute, Chalk Road, Blackboardville, USA}
\date{}

\maketitle

\begin{abstract}
	\textbf{Summary:}
	We introduce \gls{pxblat}, a Python library designed to enhance usability and efficiency in interacting with the \gls{blat}.
	\gls{pxblat} provides an intuitive \gls{api} design, allowing the incorporation of its functionality directly into Python-based bioinformatics workflows.
	Besides, it integrates seamlessly with Biopython and comes equipped with user-centric features like server readiness checks and port retry mechanisms.
	\gls{pxblat} removes the necessity for system calls and intermediate files, as well as reducing latency and data conversion overhead.
	Benchmark tests reveal \gls{pxblat} gains a \textasciitilde\SI[per-mode=symbol,round-precision=0]{20}{\percent} performance boost compared to \gls{blat} in the Python environment. \\
	\textbf{Availability and Implementation:}
	\gls{pxblat} supports Python (version 3.8+), and pre-compiled packages are released via PyPI (\url{https://pypi.org/project/pxblat/}) and Bioconda (\url{https://anaconda.org/bioconda/pxblat}).
	The source code of \gls{pxblat}  is available under the terms of an open-source MIT license and hosted on GitHub (\url{https://github.com/ylab-hi/pxblat}).
	Its documentation is available on ReadTheDocs (\url{https://pxblat.readthedocs.io/en/latest/}).
\end{abstract}

\begin{keywords}
	Software Libraries |  Sequence Analysis | BLAT
\end{keywords}

\begin{corrauthor}
	rendong.yang\at northwestern.edu
\end{corrauthor}

\section*{Introduction}\label{sec:introduction}

The rise of Python as a preferred programming language within bioinformatics is widely acknowledged as a result of its user-friendly nature, extensive libraries, and unparalleled versatility~\citep{perkel2015programming}.
A variety of libraries have been crafted to augment Python's interface, thereby amplifying the adaptability and compatibility of bioinformatics tools~\citep{putri2022analysing, cock2009biopython}.
For instance, Biopython~\citep{cock2009biopython}, a preeminent bioinformatics library, furnishes interfaces to tools like \gls{blast}~\citep{altschul1990basic} and Clustal~\citep{higgins1988clustal}.

Besides, the relentless progression of genome sequencing technologies has precipitated a dramatic surge in the availability of genomic data.
The necessity for tools to decipher and manipulate these data has become paramount within both the research and clinical domains.
\gls{blat}~\citep{kent2002blat} is a standout within the bioinformatics landscape and is recognized for its capability to swiftly conduct genome sequence alignments.
It offers a faster alternative to \gls{blast}~\citep{altschul1990basic} for aligning DNA sequences to the human genome~\citep{kent2002blat}.
Despite its widespread use and endorsement within the bioinformatics community, interfacing with BLAT can present challenges.
\gls{blat}, implemented in the C language, solely extends utilities via \glspl{cli}, rendering its integration into Python-based projects inconvenient.
Moreover, \gls{blat} triggers system calls and data conversions that cause significant overhead.
Until now, a comprehensive approach to enhancing the usability of \gls{blat} has been absent when incorporated into Python projects.

In this paper, we propose \gls{pxblat}, which empowers users to utilize \gls{blat} programmatically, facilitating its seamless integration into novel algorithms or analytical pipelines.
\gls{pxblat} serves as a bridge, bringing the high-performance capabilities of \gls{blat} into the Python environment while maintaining reproducibility.
The overarching aim of this work is to fill the observed gap by providing a Python binding library specifically tailored for \gls{blat}, addressing both its efficiency and ergonomic concerns.

\section*{Implementation}\label{sec:implementation}

The design of \gls{pxblat} adheres to the principles of readability and simplicity, aiming for an intuitive user experience that diminishes the learning curve.
Striving to curtail complexity while enhancing usability and performance, we extract the implementation of \gls{blat} from the codebase of the \gls{ucsc}, thus reducing dependencies.
We retain the original C codebase, and reimplement the utilities of \gls{blat} \(\left(\mathtt{V}37.1\right)\) including \emph{faTwoBit}, \emph{gfServer}, and \emph{gfClient} \gls{cli}, using the C\texttt{++} programming language~\citep{kent2002blat}.
Subsequently, we integrate the current C\texttt{++} code and develop \gls{pxblat} via Pybind11~\citep{pybind11}.
The approach facilitated direct interaction with the functions of the \gls{blat} without modifying its original source code, thereby preserving the integrity and performance of the \gls{blat} while broadening its capabilities.

The query result of \gls{pxblat} conforms to the \emph{QueryResult} class of Biopython~\citep{cock2009biopython}, enabling  manipulation of the query result using Biopython features~\pCref{listing:example}.
Furthermore, \gls{pxblat} eliminates the need for intermediate files, allowing all operations to be executed in memory.
This eradicates the common obstacle of data conversion into specific formats, enabling users to focus more on the core sequence alignment task.
Input and output files have been made optional, offering more flexible choices.
The use of system calls, though functional, can induce latency and create performance bottlenecks.
\gls{pxblat} reduces the need for system calls, thereby improving efficiency~\pCref{tab:performance-evaluation}.
Moreover, \gls{pxblat} allows for the retrieval of server status without manipulating log files, a process that can be troublesome in concurrent environments.
The library incorporates ergonomic features such as checking and waiting for server readiness for alignment, retrying different ports if the current one is occupied, and using an existing server if one has already been started.

We provide a variety of examples and documentation to help the user get started~\pCref{listing:example}.
Specifically, \gls{pxblat} provides \glspl{api}, including Class \emph{Server}, \emph{Client}, and other free functions, to reproduce the \gls{blat} suit.
Class \emph{Server} and \emph{Client} have the same utilities as \gls{cli} \emph{gfServer} and \emph{gfClient}, respectively, but with greater flexibility.
Free functions, for example, \emph{start\_server}, \emph{query\_server}, \emph{status\_server}, \emph{fa2twobit}, \emph{twobit2fa}, etc., serve as potential usage in different contexts.
The library has been tested and developed with \gls{ci} and \gls{cd} to ensure code quality.
It uses type annotations for public classes and functions to guarantee quality and correctness via a type checker and static analyzer.
The type annotations also make \gls{pxblat} more user-friendly in developing environments, where the function signatures can be suggested and corrected automatically.
Besides, \gls{pxblat} contains \glspl{cli} implemented by its \gls{api}.
The \gls{cli} includes completion for different shells, improving versatility.

\begin{listing}
	\inputminted[linenos, breaklines]{python}{codes/example1.py}
	\mycaption{\gls{api} Example}{The code snippet shows how to use the \gls{api} of \gls{pxblat},
		and the query result can be iterated. More code examples can be found at \url{https://pxblat.readthedocs.io/en}}
	\label{listing:example}
\end{listing}


\section*{Result}\label{sec:result}

The result and performance of \gls{pxblat} are benchmarked against \gls{blat} \(\left(\mathtt{V}37.1\right)\) using a dataset of \num[round-mode=places, round-precision=0]{1000} FASTA Files.
The dataset, which consists of five groups, is sampled from chromosome \num[round-mode=places, round-precision=0]{20} of the Human Genome (hg38), and each sample includes one sequence.
The dataset includes varying lengths of sequence to cover a range of typical usage scenarios.
The length of each sequence ranges from \num[round-mode=places, round-precision=0]{1000} to \num[round-mode=places, round-precision=0]{3000}~\pCref{suppfig:fas-len}.
We summarize the \glspl{hsp} of all hits in each sample using \gls{blat} and \gls{pxblat} and compare the results.
That ensures the correctness of \gls{pxblat}~\pCref{supptab:cmp1}.

\begin{table*}
	\centering
	\caption{Performance  Benchmarking}
	\label{tab:performance-evaluation}
	\begin{tabular}{lSSSS}
		\toprule
		Data Group      & {Samples (No.)} & {BLAT (Sec.)} & {PxBLAT (Sec.)} & {Speedup} \\
		\midrule
		benchmark/fas/1 & 200             & 61.4011       & 50.4143         & 1.2179    \\
		benchmark/fas/2 & 200             & 61.6994       & 51.0984         & 1.2075    \\
		benchmark/fas/3 & 200             & 55.3351       & 46.4404         & 1.1915    \\
		benchmark/fas/4 & 200             & 56.9760       & 47.7234         & 1.1939    \\
		benchmark/fas/5 & 200             & 54.5457       & 46.4391         & 1.1746    \\
		\bottomrule
	\end{tabular}
\end{table*}

The benchmarking is conducted on an Apple M1 Pro with macOS 13.4.1 22F82 arm64.
We use system calls to launch \gls{blat} and the \emph{time} library to measure the execution time.
Generally, \gls{pxblat} gains \textasciitilde\SI[per-mode=symbol,round-precision=0]{20}{\percent} speedup compared to \gls{blat}~\pCref{tab:performance-evaluation}.
In summary, \gls{pxblat}  offers tangible benefits according to reduced execution time and improved user experience, proving its value as an enhancement to \gls{blat}'s functionality.


\section*{Acknowledgements}\label{sec:acknowledgements}

Special thanks to the team who maintain the \gls{ucsc} codebase and users from the bioinformatics community whose valuable feedback and suggestions were pivotal in refining \gls{pxblat}'s design and functionality.
We are grateful for the generous support from our team, institution and funding bodies, which made this project possible.
We look forward to continuing to contribute to its advancement.

\section*{Conflict of interest}\label{sec:conflict-of-interest}

None declared.

\section*{Funding}\label{sec:funding}

This work was supported by the National Institute of General Medical
Sciences [R35GM142441].

\section*{Data availability}\label{sec:data-availability}


The dataset for benchmarking and testing is available at the GitHub repository {\url{https://github.com/ylab-hi/pxblat}.
The path of the dataset is \emph{benchmark/fas}.

\section*{Code availability}\label{sec:code-availability}

The \gls{pxblat}, along with the source code, is publicly available in the GitHub repository at \url{https://github.com/ylab-hi/pxblat}.
The documentation is available at ReadtheDocs \url{https://pxblat.readthedocs.io/en/latest/}.
The script for benchmarking is available at \emph{tests/test\_result.py} in the repository.


\section*{Reference}\label{sec:reference}
\bibliographystyle{bxv_abbrvnat}
\bibliography{ref.bib}
